File: backsteros-desktop/src-tauri/src/whoop.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use serde::Serialize;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WhoopSettingsStatus {
    pub connected: bool,
    pub configured: bool,
    pub email: Option<String>,
    pub reason: Option<String>,
    pub env_path: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WhoopDayResult {
    pub authenticated: bool,
    pub snapshot: Option<serde_json::Value>,
    pub error: Option<String>,
}

fn home_dir() -> Option<PathBuf> {
    std::env::var_os("HOME")
        .or_else(|| std::env::var_os("USERPROFILE"))
        .map(PathBuf::from)
}

fn totem_env_path() -> PathBuf {
    if let Ok(configured) = std::env::var("TOTEM_ENV_PATH") {
        let trimmed = configured.trim();
        if !trimmed.is_empty() {
            return PathBuf::from(trimmed);
        }
    }

    let data_dir = std::env::var("BACKSTER_DATA_DIR")
        .ok()
        .map(|value| value.trim().to_string())
        .filter(|value| !value.is_empty())
        .map(PathBuf::from)
        .or_else(|| home_dir().map(|home| home.join(".backsteros-agent")));

    data_dir
        .unwrap_or_else(|| PathBuf::from(".backsteros-agent"))
        .join("totem.env")
}

fn read_env_file(path: &Path) -> Result<std::collections::HashMap<String, String>, String> {
    if !path.exists() {
        return Ok(std::collections::HashMap::new());
    }

    let content = fs::read_to_string(path).map_err(|error| error.to_string())?;
    let mut result = std::collections::HashMap::new();

    for line in content.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let Some(separator) = trimmed.find('=') else {
            continue;
        };
        if separator == 0 {
            continue;
        }
        let key = trimmed[..separator].trim().to_string();
        let value = trimmed[separator + 1..].trim().to_string();
        if !key.is_empty() {
            result.insert(key, value);
        }
    }

    Ok(result)
}

fn mask_email(email: &str) -> String {
    let trimmed = email.trim();
    let Some(at) = trimmed.find('@') else {
        return if trimmed.is_empty() {
            String::new()
        } else {
            "••••".to_string()
        };
    };
    let local = &trimmed[..at];
    let domain = &trimmed[at..];
    let visible_len = local.len().min(2);
    let visible = &local[..visible_len];
    let dots = if local.len() > 2 { "•••" } else { "" };
    format!("{visible}{dots}{domain}")
}

fn resolve_app_dir() -> Result<PathBuf, String> {
    if let Ok(configured) = std::env::var("BACKSTEROS_APP_DIR") {
        let trimmed = configured.trim();
        if !trimmed.is_empty() {
            let path = PathBuf::from(trimmed);
            if path.is_dir() {
                return Ok(path);
            }
            return Err(format!(
                "BACKSTEROS_APP_DIR does not exist: {}",
                path.display()
            ));
        }
    }

    // src-tauri → backsteros-desktop → repo root → backsteros-app
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let candidate = manifest_dir
        .join("..")
        .join("..")
        .join("backsteros-app");
    let canonical = candidate
        .canonicalize()
        .map_err(|error| format!("Could not resolve backsteros-app: {error}"))?;
    if !canonical.is_dir() {
        return Err(format!(
            "backsteros-app not found at {}",
            canonical.display()
        ));
    }
    Ok(canonical)
}

#[tauri::command]
pub fn whoop_status() -> Result<WhoopSettingsStatus, String> {
    let env_path = totem_env_path();
    let env_path_display = env_path.display().to_string();
    let file_exists = env_path.exists();
    let env = read_env_file(&env_path)?;

    let has_refresh = env
        .get("WHOOP_COGNITO_REFRESH_TOKEN")
        .map(|value| !value.trim().is_empty())
        .unwrap_or(false);
    let has_bearer = env
        .get("WHOOP_IOS_BEARER_TOKEN")
        .map(|value| !value.trim().is_empty())
        .unwrap_or(false);
    let connected = has_refresh || has_bearer;
    let configured = connected || file_exists;
    let email = env
        .get("WHOOP_EMAIL")
        .map(|value| value.trim())
        .filter(|value| !value.is_empty())
        .map(mask_email);

    let reason = if connected {
        None
    } else if configured {
        Some(
            "totem.env exists but no valid refresh or bearer token was found.".to_string(),
        )
    } else {
        Some(
            "No totem.env file found. Configure Whoop tokens to show recovery, sleep, and strain above journal entries."
                .to_string(),
        )
    };

    Ok(WhoopSettingsStatus {
        connected,
        configured,
        email,
        reason,
        env_path: env_path_display,
    })
}

#[tauri::command]
pub fn whoop_fetch_day(date: String) -> Result<WhoopDayResult, String> {
    let status = whoop_status()?;
    if !status.connected {
        return Ok(WhoopDayResult {
            authenticated: false,
            snapshot: None,
            error: status.reason.or_else(|| {
                Some(
                    "Whoop is not connected. Add refresh or bearer tokens to totem.env."
                        .to_string(),
                )
            }),
        });
    }

    if !regex_is_iso_date(&date) {
        return Err("date must be YYYY-MM-DD".to_string());
    }

    let app_dir = resolve_app_dir()?;
    let script_path = app_dir.join("scripts").join("whoop-fetch-cli.ts");
    let tsx_bin = app_dir.join("node_modules").join(".bin").join("tsx");

    if !script_path.exists() {
        return Err(format!(
            "Whoop CLI script missing at {}",
            script_path.display()
        ));
    }

    let mut command = if tsx_bin.exists() {
        let mut cmd = Command::new(&tsx_bin);
        cmd.arg(&script_path);
        cmd
    } else {
        let mut cmd = Command::new("npx");
        cmd.args(["tsx", script_path.to_str().unwrap_or("scripts/whoop-fetch-cli.ts")]);
        cmd
    };

    command
        .arg("--date")
        .arg(&date)
        .current_dir(&app_dir)
        .env("TOTEM_ENV_PATH", totem_env_path());

    let output = command
        .output()
        .map_err(|error| format!("Failed to run Whoop CLI: {error}"))?;

    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();

    if stdout.is_empty() {
        let message = if stderr.is_empty() {
            format!(
                "Whoop CLI returned no output (exit {})",
                output.status.code().unwrap_or(-1)
            )
        } else {
            stderr
        };
        return Ok(WhoopDayResult {
            authenticated: true,
            snapshot: None,
            error: Some(message),
        });
    }

    let parsed: serde_json::Value = serde_json::from_str(&stdout).map_err(|error| {
        format!(
            "Could not parse Whoop CLI output: {error}. stdout={}",
            stdout.chars().take(200).collect::<String>()
        )
    })?;

    let ok = parsed
        .get("ok")
        .and_then(|value| value.as_bool())
        .unwrap_or(false);
    let snapshot = parsed.get("snapshot").cloned().filter(|value| !value.is_null());
    let error = parsed
        .get("error")
        .and_then(|value| value.as_str())
        .map(str::to_string)
        .or_else(|| {
            if !ok && !stderr.is_empty() {
                Some(stderr)
            } else {
                None
            }
        });

    Ok(WhoopDayResult {
        authenticated: true,
        snapshot: if ok { snapshot } else { None },
        error: if ok {
            None
        } else {
            error.or_else(|| Some("Whoop fetch failed".to_string()))
        },
    })
}

fn regex_is_iso_date(value: &str) -> bool {
    if value.len() != 10 {
        return false;
    }
    let bytes = value.as_bytes();
    bytes[4] == b'-'
        && bytes[7] == b'-'
        && bytes[..4].iter().all(u8::is_ascii_digit)
        && bytes[5..7].iter().all(u8::is_ascii_digit)
        && bytes[8..].iter().all(u8::is_ascii_digit)
}

File: core/server/src/services/whoop.ts
import { createRequire } from "node:module";
import { existsSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { pathToFileURL } from "node:url";

import type { WhoopDayResult, WhoopSnapshot, WhoopSettingsStatus } from "@backsteros/contracts";

const require = createRequire(import.meta.url);

const WHOOP_DAY_CACHE_TTL_MS = 120_000;
const whoopDayCache = new Map<
  string,
  { writtenAt: number; result: WhoopDayResult }
>();

type TotemWhoopModules = {
  TokenManager: new (config: {
    email: string;
    accessToken: string;
    refreshToken: string;
    envPath: string;
  }) => { getToken: () => Promise<string> };
  WhoopClient: new (config: {
    getToken: () => Promise<string>;
  }) => {
    get: (path: string, query?: Record<string, string>) => Promise<unknown>;
  };
  projectToday: (input: {
    home: unknown;
    sleep: unknown;
    recovery: unknown;
    state: unknown;
    date: string;
  }) => unknown;
};

let totemModulesPromise: Promise<TotemWhoopModules> | null = null;

function getTotemEnvPath(): string {
  const configured = process.env.TOTEM_ENV_PATH?.trim();
  if (configured) return configured;
  const dataDir =
    process.env.BACKSTER_DATA_DIR?.trim() ||
    join(homedir(), ".backsteros-agent");
  return join(dataDir, "totem.env");
}

function readEnvFile(path: string): Record<string, string> {
  if (!existsSync(path)) return {};
  const result: Record<string, string> = {};
  for (const line of readFileSync(path, "utf8").split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const separator = trimmed.indexOf("=");
    if (separator <= 0) continue;
    result[trimmed.slice(0, separator).trim()] =
      trimmed.slice(separator + 1).trim();
  }
  return result;
}

function maskEmail(email: string): string {
  const trimmed = email.trim();
  const at = trimmed.indexOf("@");
  if (at < 0) return trimmed ? "••••" : "";
  const local = trimmed.slice(0, at);
  const domain = trimmed.slice(at);
  const visibleLen = Math.min(2, local.length);
  const visible = local.slice(0, visibleLen);
  const dots = local.length > 2 ? "•••" : "";
  return `${visible}${dots}${domain}`;
}

function getWhoopEnv(): Record<string, string> {
  const fileEnv = readEnvFile(getTotemEnvPath());
  const keys = [
    "WHOOP_EMAIL",
    "WHOOP_IOS_BEARER_TOKEN",
    "WHOOP_COGNITO_REFRESH_TOKEN",
  ] as const;
  const env: Record<string, string> = {};
  for (const key of keys) {
    const value = (process.env[key]?.trim() || fileEnv[key]?.trim() || "");
    if (value) env[key] = value;
  }
  return env;
}

function formatDateInTimezone(timezone: string, date = new Date()): string {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: timezone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(date);
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object"
    ? (value as Record<string, unknown>)
    : null;
}

function asNumber(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function asInt(value: unknown): number | null {
  const num = asNumber(value);
  return num == null ? null : Math.trunc(num);
}

function asRecoveryState(
  value: unknown,
): "GREEN" | "YELLOW" | "RED" | null {
  return value === "GREEN" || value === "YELLOW" || value === "RED"
    ? value
    : null;
}

function formatDuration(ms: number | null | undefined): string | undefined {
  if (ms == null) return undefined;
  const totalMinutes = Math.round(ms / 60_000);
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  if (hours <= 0) return `${minutes}m`;
  if (minutes <= 0) return `${hours}h`;
  return `${hours}h ${minutes}m`;
}

function parseSleepStages(value: unknown) {
  const stages = asRecord(value);
  if (!stages) return undefined;
  return {
    remMs: asInt(stages.rem_ms),
    lightMs: asInt(stages.light_ms),
    swsMs: asInt(stages.sws_ms),
    wakeMs: asInt(stages.wake_ms),
  };
}

function parseTodaySnapshot(record: Record<string, unknown>): WhoopSnapshot | null {
  const date = typeof record.date === "string" ? record.date : undefined;
  const recovery = asRecord(record.recovery);
  const sleep = asRecord(record.sleep);
  const strain = asRecord(record.strain);
  if (!date || !recovery || !sleep || !strain) return null;

  return {
    id: `whoop-${date}`,
    date,
    recoveryScore: asNumber(recovery.score),
    recoveryState: asRecoveryState(recovery.state),
    hrvMs: asNumber(recovery.hrv_ms),
    rhrBpm: asNumber(recovery.rhr_bpm),
    sleepPerformance: asNumber(sleep.performance_pct),
    sleepDuration: formatDuration(asInt(sleep.total_sleep_ms)),
    strainScore: asNumber(strain.score),
    workoutsCount: asInt(strain.workouts_count) ?? 0,
    sleepStartedAt:
      typeof sleep.started_at === "string" ? sleep.started_at : null,
    sleepEndedAt: typeof sleep.ended_at === "string" ? sleep.ended_at : null,
    timeInBed: formatDuration(asInt(sleep.time_in_bed_ms)),
    sleepEfficiencyPct: asNumber(sleep.efficiency_pct),
    sleepStages: parseSleepStages(sleep.stages),
  };
}

async function loadTotemModules(): Promise<TotemWhoopModules> {
  if (!totemModulesPromise) {
    totemModulesPromise = (async () => {
      let totemPackageJson: string;
      try {
        totemPackageJson = require.resolve("@briangaoo/totem/package.json");
      } catch {
        throw new Error(
          "Missing @briangaoo/totem. Run `pnpm install` from the repo root.",
        );
      }

      const totemRoot = dirname(totemPackageJson);
      const cognitoDist = join(totemRoot, "dist", "whoop", "cognito.js");
      if (!existsSync(cognitoDist)) {
        throw new Error(
          "Whoop totem dist is not built. Run `node scripts/whoop/build-totem.mjs` in core/server.",
        );
      }

      async function load(...parts: string[]) {
        const href = pathToFileURL(join(totemRoot, "dist", ...parts)).href;
        return import(href);
      }

      const [tokenManagerModule, clientModule, todayModule] = await Promise.all([
        load("whoop", "token_manager.js"),
        load("whoop", "client.js"),
        load("projections", "today.js"),
      ]);

      return {
        TokenManager: tokenManagerModule.TokenManager as TotemWhoopModules["TokenManager"],
        WhoopClient: clientModule.WhoopClient as TotemWhoopModules["WhoopClient"],
        projectToday: todayModule.projectToday as TotemWhoopModules["projectToday"],
      };
    })();
  }
  return totemModulesPromise;
}

export function getWhoopSettingsStatus(): WhoopSettingsStatus {
  const envPath = getTotemEnvPath();
  const fileExists = existsSync(envPath);
  const env = getWhoopEnv();
  const connected = Boolean(
    env.WHOOP_COGNITO_REFRESH_TOKEN || env.WHOOP_IOS_BEARER_TOKEN,
  );
  const configured = connected || fileExists;
  const email = env.WHOOP_EMAIL ? maskEmail(env.WHOOP_EMAIL) : null;

  let reason: string | null = null;
  if (!connected) {
    reason = configured
      ? "totem.env exists but no valid refresh or bearer token was found."
      : "No totem.env file found. Configure Whoop tokens to show recovery, sleep, and strain above journal entries.";
  }

  return {
    connected,
    configured,
    email,
    reason,
    envPath,
  };
}

async function fetchWhoopDaySnapshotUncached(
  date: string,
): Promise<WhoopDayResult> {
  const status = getWhoopSettingsStatus();
  if (!status.connected) {
    return {
      authenticated: false,
      snapshot: null,
      error:
        status.reason ??
        "Whoop is not connected. Add refresh or bearer tokens to totem.env.",
    };
  }

  const env = getWhoopEnv();
  const email = env.WHOOP_EMAIL;
  const accessToken = env.WHOOP_IOS_BEARER_TOKEN;
  const refreshToken = env.WHOOP_COGNITO_REFRESH_TOKEN;
  if (!email || !accessToken || !refreshToken) {
    return {
      authenticated: false,
      snapshot: null,
      error: "Whoop tokens are incomplete in totem.env",
    };
  }

  try {
    const { TokenManager, WhoopClient, projectToday } = await loadTotemModules();
    const timezone =
      process.env.WHOOP_TIMEZONE?.trim() || "Europe/Amsterdam";
    const resolvedDate = date || formatDateInTimezone(timezone);

    const tokenManager = new TokenManager({
      email,
      accessToken,
      refreshToken,
      envPath: getTotemEnvPath(),
    });
    const client = new WhoopClient({
      getToken: () => tokenManager.getToken(),
    });

    const [home, sleep, recovery, state] = await Promise.all([
      client.get("/home-service/v1/home", { date: resolvedDate }),
      client.get("/developer/v2/activity/sleep", { limit: "5" }).catch(() => null),
      client.get("/developer/v2/recovery", { limit: "5" }).catch(() => null),
      client.get("/activities-service/v1/user-state").catch(() => null),
    ]);

    const projected = projectToday({
      home,
      sleep,
      recovery,
      state,
      date: resolvedDate,
    });
    const snapshot = parseTodaySnapshot(asRecord(projected) ?? {});

    return {
      authenticated: true,
      snapshot,
      error: snapshot ? null : "No Whoop snapshot for this date.",
    };
  } catch (error) {
    return {
      authenticated: true,
      snapshot: null,
      error:
        error instanceof Error ? error.message : "Failed to load Whoop data",
    };
  }
}

export async function fetchWhoopDaySnapshot(
  date: string,
): Promise<WhoopDayResult> {
  const cached = whoopDayCache.get(date);
  if (
    cached &&
    Date.now() - cached.writtenAt < WHOOP_DAY_CACHE_TTL_MS &&
    cached.result.authenticated &&
    cached.result.snapshot
  ) {
    return cached.result;
  }

  const result = await fetchWhoopDaySnapshotUncached(date);
  if (result.authenticated && result.snapshot) {
    whoopDayCache.set(date, { writtenAt: Date.now(), result });
  } else {
    whoopDayCache.delete(date);
  }
  return result;
}
